src/web100srv.c

```diff
@@ -1527,647 +1326,364 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
+  if (!queue) {
+    max_simultaneous_tests = 1;
+  } else {
+    max_simultaneous_tests = max_clients;
+  }
+  queue_position = 0;
+  for (tmp_ptr = *head; tmp_ptr != NULL; tmp_ptr = tmp_ptr->next) {
+    if (!tmp_ptr->running) {
+      if (running_count < max_simultaneous_tests) {
+        tmp_ptr->running = 1;
+        child_message = SRV_QUEUE_TEST_STARTS_NOW;
+        write(tmp_ptr->pipe, &child_message, sizeof(ParentChildSignal));
+        running_count++;
       } else {
-        log_println(4, "New request has arrived, adding request to queue list");
-        while ((retcode = sem_wait(&ndtq)) == -1 && errno == EINTR) {
-          log_println(6,
-                      "Waiting for ndtq semaphore to free, adding "
-                      "new client 2");
-          continue;
-        }
-        log_println(6, "Adding new client to fifo queue - semaphore locked");
-        tmp_ptr = head_ptr;
-        while (tmp_ptr->next != NULL) tmp_ptr = tmp_ptr->next;
-        tmp_ptr->next = new_child;
-        sem_post(&ndtq);
-        log_println(6, "Free'd ndtq semaphore lock - 2");
-
-        if (get_debuglvl() > 3) {
-          log_println(4, "Walking scheduling queue");
-          tmp_ptr = head_ptr;
-          while (tmp_ptr != NULL) {
-            log_println(4, "\tChild %d, host: %s [%s], next=0x%x", tmp_ptr->pid,
-                        tmp_ptr->host, tmp_ptr->addr, (uintptr_t)tmp_ptr->next);
-            if (tmp_ptr->next == NULL) break;
-            tmp_ptr = tmp_ptr->next;
-          }
-        }
-      }
-
-      // At this point send a message to the client via the ctlsockfd
-      // saying that N clients are waiting in the queue & testing will
-      // begin within Nx60 seconds.  Only if (waiting > 0)
-      // Clients who leave will be handled in the run_test routine.
-      if ((multiple == 0) && (waiting > 1)) {
-        child_signal = waiting - 1;
-        log_println(3,
-                    "%d clients waiting, telling client (%d) testing will "
-                    "begin within %d minutes",
-                    child_signal, tmp_ptr->pid, child_signal);
-        write(child_pipe[1], &child_signal, sizeof(int));
-        continue;
+        // Assuming we can service max_simultaneous_tests per minute, then the
+        // amount of time that the child should be expected to wait is equal to
+        // the number of tests ahead of it, divided by the number of tests we
+        // can service per minute.  If the children are receiving too many
+        // updates, don't change the update frequency here.  Instead, change
+        // how the children respond to updates in child_process().
+        child_message = max(queue_position / max_simultaneous_tests, 1);
+        write(tmp_ptr->pipe, &child_message, sizeof(ParentChildSignal));
+        child_message = SRV_QUEUE_HEARTBEAT;
+        write(tmp_ptr->pipe, &child_message, sizeof(ParentChildSignal));
       }
+    }
+    queue_position++;
+  }
+}
```

Comment:
Can we break this up?

1: `purge_dead_clients()`: 1557-1585
2: `count_running_clients()`: 1587-1595
3: `update_waiting_children()`: 1597-1626

2 & 3 I can see merging, but 1 & 3 are very distinct.